### PR TITLE
feat(jdl): itinerary job, and steps states

### PR DESCRIPTION
**Pull request by kepkap**
this is DRAFT PR

JDL initial implementation

1. Itinerary individual steps lay nicely on finite states, and can be modelled as state machines (FSM).
2. itinerary flow itself is dynamic, and looks more like a workflow in a sense that it is not event driven, but next step is driven by completion of previous step. hence workflow-like.

NB! the flow draft implementation assumes itinerary job step are sequential in execution.

### CODE REVIEW

File: core/model/job/itinerary/itinerary.go [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+type Stepf func()
+
+type Passenger struct {
+	ID   string
+	Name string
+}
+
+type Itinerary struct {
+	ID        string
+	StartTime string
```
Comment (vlkolmakov) [inline core/model/job/itinerary/itinerary.go:12]: Why `StartTime` is a string, not a `time.Time`?

**Comment by vlkolmakov** [comment]
**General Note**

Using linked lists might be theoretically correct, but practically it's gonna be a mess compared to a plain []slice.

The only benefit of a linked list is fast insertions/deletions in the middle. Since we're not going to do ins/dels hundred of times per sec, slice operations are sufficient.

**Moreover, we loose:**
  - range iteration
  - fast and painless serialization to JSON/YAML
  - fast database operations

File: core/model/job/itinerary/flow.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+type StepType int
+
+const (
+	goToLocation StepType = iota
+	pickUpPassengers
+	dropOffPassengers
+)
+
+type step struct {
```
Comment (vlkolmakov) [inline core/model/job/itinerary/flow.go:13]: `step` should be public

File: core/model/job/itinerary/flow.go [hunk 3]
```diff
@@ -0,0 +1,89 @@
+package itinerary
+
+import "fmt"
+
+type StepType int
```
Comment (vlkolmakov) [inline core/model/job/itinerary/flow.go:5]: `StepType` should implement `String() string`

**Comment by kepkap** [comment]
@vlkolmakov please delay review until PR is not DRAFT.

File: core/model/job/itinerary/flow.go [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+	goToLocation StepType = iota
+	pickUpPassengers
+	dropOffPassengers
+)
+
+type step struct {
+	prev     *step
+	next     *step
+	id       string
+	stepType StepType
```
[the remark this answers is not in the document]
Comment (kevinmichaelchen) [reply core/model/job/itinerary/flow.go:17]: Only if it's being used outside of the `itinerary` package, right? Otherwise it's fine to not export it?

**Comment by kepkap** [comment]
will reopen new PR, over another diverged branch. this branch is obsolete.

**Pull request closed without merging**